src/main/client_context.cpp

```diff
@@ -73,8 +74,39 @@ struct ActiveQueryContext {
 	BaseQueryResult *open_result = nullptr;
 };
 
+#ifdef DEBUG
+struct DebugClientContextState : public ClientContextState {
+	~DebugClientContextState() override {
+		D_ASSERT(active_transaction == 0);
+		D_ASSERT(active_query == 0);
+	}
+
+	idx_t active_transaction = 0;
+	idx_t active_query = 0;
+
+	void QueryBegin(ClientContext &context) override {
+		active_query++;
+	}
+	void QueryEnd(ClientContext &context) override {
```

Comment:
These lifecycle hooks are such a great way to check that DuckDB is behaving correctly. For example, we could assert here that ClientContext no longer has an `active_query` / `active_query.executor`.

Reply:
The issue with that is that `QueryEnd` is currently called before we reset `active_query`, so that wouldn't work, but yes agreed it's a useful tool

Reply:
Understood. How do you feel about inverting the calls to `active_query->executor->CancelTasks();` and `QueryEnd()`?

Unless execution has stopped, a query is not actually ended.

Reply:
Sure, fine by me